collect_app/src/test/java/org/odk/collect/android/formentry/QuitFormDialogFragmentTest.java

```diff
@@ -0,0 +1,96 @@
+package org.odk.collect.android.formentry;
+
+import android.content.DialogInterface;
+
+import androidx.appcompat.app.AlertDialog;
+import androidx.fragment.app.FragmentActivity;
+import androidx.fragment.app.FragmentManager;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.odk.collect.android.R;
+import org.odk.collect.android.formentry.saving.FormSaveViewModel;
+import org.odk.collect.android.support.RobolectricHelpers;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.shadows.ShadowDialog;
+
+import static android.view.View.GONE;
+import static android.view.View.VISIBLE;
+import static junit.framework.TestCase.assertTrue;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.odk.collect.android.support.RobolectricHelpers.mockViewModelProvider;
+import static org.robolectric.Shadows.shadowOf;
+
+@RunWith(RobolectricTestRunner.class)
+public class QuitFormDialogFragmentTest {
+
+    private FragmentActivity activity;
+    private FragmentManager fragmentManager;
+    private QuitFormDialogFragment dialogFragment;
+
+    private FormSaveViewModel viewModel;
+
+    @Before
+    public void setup() {
+        activity = RobolectricHelpers.createThemedActivity(FragmentActivity.class);
+        fragmentManager = activity.getSupportFragmentManager();
+        dialogFragment = new QuitFormDialogFragment();
+
+        viewModel = mockViewModelProvider(activity, FormSaveViewModel.class).get(FormSaveViewModel.class);
+    }
+
+    @Test
+    public void shouldShowCorrectView() {
+        dialogFragment.show(fragmentManager, "TAG");
+        AlertDialog dialog = (AlertDialog) dialogFragment.getDialog();
+
+        assertNotNull(dialog);
+        assertThat(dialog.getButton(DialogInterface.BUTTON_NEGATIVE).getVisibility(), equalTo(GONE));
+        assertThat(dialog.getButton(DialogInterface.BUTTON_POSITIVE).getVisibility(), equalTo(VISIBLE));
+        assertThat(dialog.getButton(DialogInterface.BUTTON_POSITIVE).getText(),
+                equalTo(activity.getString(R.string.do_not_exit)));
+
+        String title = dialogFragment.getTitle(viewModel);
```

Comment:
I'm confused by your use of `getTitle` and `setTitle` here. Could you explain why you chose to create a public `getTitle` and `setTitle`?

Reply:
I created the public methods so that I can use them in the `QuitFormDialogFragmentTest`.

Reply:
Right. It's generally best for us to avoid creating methods just for tests. Our tests should act like "real" clients of whatever code they are testing and so shouldn't have access to methods or fields we wouldn't want them to have access to. Feeling like you need to make a method public for a test can be a nice hint that something is probably off about the design of the object you're testing or that you're testing the wrong thing!

Here you shouldn't have to do anything but `show` the `DialogFragment` to have the title set right? I think you actually end up with a test that you don't need as there is no mutable state that needs needs to carried over from one instance of the fragment to the other: the title always just comes from the view model.